### PR TITLE
Fix occasional IOMCU internal errors on startup

### PR DESCRIPTION
**Pull request by andyp1per**
- allow DIRECT_PWM pages to be smaller than max channels
- correct some over-eager register clearing in the global interrupt handler (NFC)
- only sent TX events when using shared DMA (NFC)
- zero out rx packet code and size to prevent errors with spurious callbacks
- add a comment and check for offsets that are codes

### CODE REVIEW

**Comment by tridge** [comment]
@andyp1per when switching between 4.4 and this PR I get a lot of IOMCU startup msgs:
image

File: libraries/AP_IOMCU/iofirmware/iofirmware.cpp [hunk 1]
```diff
@@ -942,7 +961,7 @@ bool AP_IOMCU_FW::handle_code_write()
             // no input when override is active
             break;
         }
-        if (rx_io_packet.count != sizeof(reg_direct_pwm.pwm)/2) {
+        if (rx_io_packet.count > sizeof(reg_direct_pwm.pwm)/2) {
```
Comment (tridge) [inline libraries/AP_IOMCU/iofirmware/iofirmware.cpp:964]: is AP_IOMCU ever sending less than a full set? if so, why?

**Review by tridge — COMMENTED** [review]
i'm having troubles testing this PR, see discussion in chibios channel

**Comment by tridge** [comment]
2nd problem is if I load this PR, then switch to 4.4.4, it fails to update IO fw. Fix is to hold down safety switch while powering on, but most users don't have a GPS with a safety switch these days

**Comment by tridge** [comment]
reproduction:
```
cd Copter-4.4; rm -rf build; ./waf configure --board CubeOrange -g && nice ./waf copter --upload
cd ../THISPR; rm -rf build; ./waf configure --board CubeOrange -g; nice ./waf plane --upload
```
gets stuck in IOMCU fw update, not fixed with power cycle

File: libraries/AP_IOMCU/iofirmware/iofirmware.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/iofirmware.cpp:964]: Yes - https://github.com/ArduPilot/ardupilot/blob/5c1ed968b18ea196b8376cdd6cc7aeae9b3085c4/libraries/AP_IOMCU/AP_IOMCU.cpp#L370 - don't know why as I didn't write this code.

Comment (andyp1per) [reply libraries/AP_IOMCU/iofirmware/iofirmware.cpp:964]: But I verified on the IOMCU side that this was happening

**Comment by andyp1per** [comment]
@tridge I rebased and rebuilt to before the millis() optimizations went in and the problem goes away - ergo I believe those changes to be responsible

**Comment by tridge** [comment]
@andyp1per this change helps: https://github.com/ArduPilot/ardupilot/pull/25826

**Review by tridge — APPROVED** [review]

**Pull request merged**